rp2040-hal/src/i2c.rs

```diff
@@ -1,3 +1,384 @@
-// TODO
+
+use crate::gpio::pin::bank0::{
+    Gpio0, Gpio1, Gpio10, Gpio11, Gpio12, Gpio13, Gpio14, Gpio15, Gpio16, Gpio17, Gpio18, Gpio19,
+    Gpio2, Gpio20, Gpio21, Gpio26, Gpio27, Gpio3, Gpio4, Gpio5, Gpio6, Gpio7, Gpio8, Gpio9,
+};
+use embedded_time::rate::Hertz;
+use hal::blocking::i2c::{Write, WriteRead};
+use rp2040_pac::{I2C0, I2C1, RESETS};
+
+/// I2C error
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum Error {
+    /// I2c abort with error
+    Abort(u32),
+}
+
+#[doc(hidden)]
+mod sealed {
+
+    use crate::gpio::pin::bank0::{
+        Gpio0, Gpio1, Gpio10, Gpio11, Gpio12, Gpio13, Gpio14, Gpio15, Gpio16, Gpio17, Gpio18,
+        Gpio19, Gpio2, Gpio20, Gpio21, Gpio26, Gpio27, Gpio3, Gpio4, Gpio5, Gpio6, Gpio7, Gpio8,
+        Gpio9,
+    };
+
+    pub trait Sealed {}
+    impl Sealed for Gpio0 {}
+    impl Sealed for Gpio1 {}
+    impl Sealed for Gpio2 {}
+    impl Sealed for Gpio3 {}
+    impl Sealed for Gpio4 {}
+    impl Sealed for Gpio5 {}
+    impl Sealed for Gpio6 {}
+    impl Sealed for Gpio7 {}
+    impl Sealed for Gpio8 {}
+    impl Sealed for Gpio9 {}
+    impl Sealed for Gpio10 {}
+    impl Sealed for Gpio11 {}
+    impl Sealed for Gpio12 {}
+    impl Sealed for Gpio13 {}
+    impl Sealed for Gpio14 {}
+    impl Sealed for Gpio15 {}
+    impl Sealed for Gpio16 {}
+    impl Sealed for Gpio17 {}
+    impl Sealed for Gpio18 {}
+    impl Sealed for Gpio19 {}
+    impl Sealed for Gpio20 {}
+    impl Sealed for Gpio21 {}
+    impl Sealed for Gpio26 {}
+    impl Sealed for Gpio27 {}
+}
+
+/// SCL pin
+pub trait SclPin<I2C>: sealed::Sealed {}
+
+/// SDA pin
+pub trait SdaPin<I2C>: sealed::Sealed {}
+
+impl SdaPin<I2C0> for Gpio0 {}
+impl SclPin<I2C0> for Gpio1 {}
+
+impl SdaPin<I2C1> for Gpio2 {}
+impl SclPin<I2C1> for Gpio3 {}
+
+impl SdaPin<I2C0> for Gpio4 {}
+impl SclPin<I2C0> for Gpio5 {}
+
+impl SdaPin<I2C1> for Gpio6 {}
+impl SclPin<I2C1> for Gpio7 {}
+
+impl SdaPin<I2C0> for Gpio8 {}
+impl SclPin<I2C0> for Gpio9 {}
+
+impl SdaPin<I2C1> for Gpio10 {}
+impl SclPin<I2C1> for Gpio11 {}
+
+impl SdaPin<I2C0> for Gpio12 {}
+impl SclPin<I2C0> for Gpio13 {}
+
+impl SdaPin<I2C1> for Gpio14 {}
+impl SclPin<I2C1> for Gpio15 {}
+
+impl SdaPin<I2C0> for Gpio16 {}
+impl SclPin<I2C0> for Gpio17 {}
+
+impl SdaPin<I2C1> for Gpio18 {}
+impl SclPin<I2C1> for Gpio19 {}
+
+impl SdaPin<I2C0> for Gpio20 {}
+impl SclPin<I2C0> for Gpio21 {}
+
+impl SdaPin<I2C1> for Gpio26 {}
+impl SclPin<I2C1> for Gpio27 {}
+
+/// I2C peripheral operating in master mode
+pub struct I2c<I2C, PINS> {
+    i2c: I2C,
+    pins: PINS,
+}
+
+fn i2c_reserved_addr(addr: u8) -> bool {
+    (addr & 0x78) == 0 || (addr & 0x78) == 0x78
+}
+
+macro_rules! hal {
+    ($($I2CX:ident: ($i2cX:ident),)+) => {
+        $(
+            impl<SCL, SDA> I2c<$I2CX, (SCL, SDA)> {
+                /// Configures the I2C peripheral to work in master mode
+                pub fn $i2cX<F>(i2c: $I2CX, pins: (SCL, SDA), freq: F) -> Self
```

Comment:
`pins` accepts two `PinId`s, not two `Pin`s -- this should be something like `Pin<SCL,FunctionI2C>`, which also handles making sure the pin is in the correct mode for being used as an i2c pin

Reply:
Also probably take this as two arguments instead of as a tuple. You can always continue to use a tuple internally.

Reply:
Fixed!